### PR TITLE
Fix index mapping bugs in contains, array_position and subscript Presto functions

### LINKED ISSUE #6346: contains(), subscript(), array_position() fail with Index is too large
### Bug description

Found by running Fuzzer (#6312) with "--only contains"

```
I20230830 06:53:52.203747 4859347 ExpressionFuzzer.cpp:1313] ==============================> Started iteration 6 (seed: 3597353198)
I20230830 06:53:52.411063 4859347 ExpressionVerifier.cpp:86] Executing expression 0 : contains("c0"["row_field1"],"c1"["row_field0"]["row_field2"]["row_field0"])
I20230830 06:53:52.411126 4859347 ExpressionVerifier.cpp:86] Executing expression 1 : contains("c0"["row_field1"],"c1"["row_field0"]["row_field2"]["row_field0"])
E20230830 06:53:52.416990 4859347 Exceptions.h:69] Line: /Users/mbasmanova/cpp/velox-1/./velox/vector/BaseVector.h:145, Function:isNullAt, Expression: idx < length_ (94 vs. 94) Index is too large, Source: RUNTIME, ErrorCode: INVALID_STATE
E20230830 06:53:52.417413 4859347 ExpressionVerifier.cpp:176] Common eval: Exceptions other than VeloxUserError are not allowed.
I20230830 06:53:52.417447 4859347 ExpressionVerifier.cpp:252] Skipping persistence because repro path is empty.
libc++abi: terminating due to uncaught exception of type facebook::velox::VeloxRuntimeError: Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: (94 vs. 94) Index is too large
Retriable: False
Expression: idx < length_
Context: contains((c0).row_field1, (((c1).row_field0).row_field2).row_field0)
Top-Level Context: Same as context.
Function: isNullAt
File: /Users/mbasmanova/cpp/velox-1/./velox/vector/BaseVector.h
Line: 145
```

### System information

n/a

### Relevant logs

_No response_

### PR DESCRIPTION
**Pull request by duanmeng**
Fixes #6346

The cardinality of a top-level vector, such as`ConstantVector<ComplextType>`, may be greater than its base vector, so we should use elementsDecoded and its indices to process each element.

### CODE REVIEW

**Review by mbasmanova — COMMENTED** [review]
Let's remove these functions from the skip list in velox/expression/tests/ExpressionFuzzerTest.cpp and verify that fuzzer no longer fails.

**Comment by duanmeng** [comment]
> Let's remove these functions from the skip list in velox/expression/tests/ExpressionFuzzerTest.cpp and verify that fuzzer no longer fails.

@mbasmanova I've removed `contains`, `subscript`, and `array_position` and the fuzzer test looks fine.

**Comment by mbasmanova** [comment]
> I've removed contains, subscript, and array_position and the fuzzer test looks fine.

Thanks. Let's update PR title and description to mention subscript as well?

> the fuzzer test looks fine.

Would you share the command line you used to run the fuzzer? Did this command fail without the fixes? Let's run fuzzer for 40+ minutes to ensure there are no issues remaining.

**Comment by duanmeng** [comment]
> Would you share the command line you used to run the fuzzer? Did this command fail without the fixes? 

I used to run and reproduce the issue in my ide by running `velox_expression_fuzzer_test` with parameters `--enable_dereference --seed 1 --enable_variadic_signatures --velox_fuzzer_enable_complex_types --lazy_vector_generation_ratio 0.2 --velox_fuzzer_enable_column_reuse --velox_fuzzer_enable_expression_reuse --max_expression_trees_per_step 2 --retry_with_try --enable_dereference --duration_sec 60 --logtostderr=1 --minloglevel=0 --repro_persist_path=./tmp/fuzzer_repro --only contains`.
After removing these functions from the skip list, I rerun the test with `--only contains,subscript,array_position`, and the fuzzer test of the ci is green.

>Let's run fuzzer for 40+ minutes to ensure there are no issues remaining.
OK, will run locally.

File: velox/functions/lib/SubscriptUtil.h [hunk 1]
```diff
@@ -433,7 +434,8 @@ class SubscriptImpl : public exec::Subscript {
       bool found = false;
       auto searchIndex = searchIndices[row];
       for (auto i = 0; i < size; i++) {
-        if (mapKeysBase->equalValueAt(searchBase, offset + i, searchIndex)) {
+        if (mapKeysBase->equalValueAt(
```
Comment (laithsakka) [inline velox/functions/lib/SubscriptUtil.h:437]: alternatively we can use mapKeys and offset + i.  and delete lines 416 and 417. 
  for (auto i = 0; i < size; i++) {
        if (mapKeys->equalValueAt(searchBase, offset + i, searchIndex)) 
        
 i assume this was done here like this for performance reasons? but its a virtual call anyway i will let you decide

Comment (laithsakka) [reply velox/functions/lib/SubscriptUtil.h:437]: same apply to the indexArg, not sure why we are decoding it 
we could have done instead:
if (mapKeys->equalValueAt(indexArg.get(), offset + i, row))

Like we did not need to decode maps keys nor indices, its not clear to me that its  a saving

File: velox/functions/prestosql/ArrayContains.cpp [hunk 2]
```diff
@@ -104,10 +104,10 @@ void applyComplexType(
 
     auto searchIndex = searchIndices[row];
     for (auto i = 0; i < size; i++) {
-      if (elementsBase->isNullAt(offset + i)) {
+      if (elementsDecoded.isNullAt(offset + i)) {
         foundNull = true;
       } else if (elementsBase->equalValueAt(
```
Comment (laithsakka) [inline velox/functions/prestosql/ArrayContains.cpp:109]: same comments as above not sure why did we decode the elements.

File: velox/functions/prestosql/tests/ArrayContainsTest.cpp [hunk 3]
```diff
@@ -26,14 +27,24 @@ using namespace facebook::velox::functions::test;
...
 
-class ArrayContainsTest : public FunctionBaseTest {};
-
-TEST_F(ArrayContainsTest, integerNoNulls) {
-  auto arrayVector = makeArrayVector<int64_t>(
-      {{1, 2, 3, 4}, {3, 4, 5}, {}, {5, 6, 7, 8, 9}, {7}, {10, 9, 8, 7}});
+class ArrayContainsTest : public FunctionBaseTest {
+ public:
+  void testContains(
+      ArrayVectorPtr arrayVector,
```
Comment (laithsakka) [inline velox/functions/prestosql/tests/ArrayContainsTest.cpp:33]: const&

File: velox/functions/prestosql/tests/ArrayContainsTest.cpp [hunk 4]
```diff
@@ -26,14 +27,24 @@ using namespace facebook::velox::functions::test;
...
+      std::vector<int64_t> search,
```
Comment (laithsakka) [inline velox/functions/prestosql/tests/ArrayContainsTest.cpp:34]: nit: ditto and remove the move, i found that simpler to follow.

File: velox/functions/prestosql/tests/ArrayContainsTest.cpp [hunk 5]
```diff
@@ -26,14 +27,24 @@ using namespace facebook::velox::functions::test;
...
+      const std::vector<std::optional<bool>>& expected) {
+    auto constSearch = BaseVector::wrapInConstant(
+        1, 0, makeArrayVector<int64_t>({std::move(search)}));
+    auto result = evaluate<SimpleVector<bool>>(
```
Comment (laithsakka) [inline velox/functions/prestosql/tests/ArrayContainsTest.cpp:38]: nit: I have seen this form a lot, whats the purpose of it evaluate<SimpleVector<bool>> -> evaluate(..)?
anyway we assert vector equivalence after that.

File: velox/functions/prestosql/tests/ArrayContainsTest.cpp [hunk 6]
```diff
@@ -207,21 +200,12 @@ TEST_F(ArrayContainsTest, booleanWithNulls) {
...
-            makeConstant(search, arrayVector->size()),
-        }));
-
-    assertEqualVectors(makeNullableFlatVector<bool>(expected), result);
-  };
-
-  testContains(true, {true, true, std::nullopt, false, true, false});
-  testContains(false, {true, false, true, false, true, true});
-  testContains(
+  testContains<bool>(
```
Comment (laithsakka) [inline velox/functions/prestosql/tests/ArrayContainsTest.cpp:203]: curious why did you need to pass the type?
is it generating ambiguous calls?
if so shall we instead have 
testContains
testContainsCosntantKey

**Comment by duanmeng** [comment]
@mbasmanova I used the following command to run the fuzzer and all passed.

`_build/debug/velox/expression/tests/velox_expression_fuzzer_test --enable_dereference --seed 0 --enable_variadic_signatures --velox_fuzzer_enable_complex_types --lazy_vector_generation_ratio 0.2 --velox_fuzzer_enable_column_reuse --velox_fuzzer_enable_expression_reuse --max_expression_trees_per_step 2 --retry_with_try --enable_dereference --duration_sec 2400 --logtostderr=1 --minloglevel=0 --repro_persist_path=./tmp/fuzzer_repro 2>&1 | tee tmp/test.log`

```C++
$ head test.log
W20230906 02:07:11.387956 537432 ExpressionFuzzer.cpp:457] Skipping unsupported signature with generic: array_frequency(array(hugeint)) -> map(hugeint,integer)
W20230906 02:07:11.394208 537432 ExpressionFuzzer.cpp:457] Skipping unsupported signature with generic: combinations(array(hugeint),bigint) -> array(array(hugeint))
W20230906 02:07:11.395435 537432 ExpressionFuzzer.cpp:457] Skipping unsupported signature with generic: array_union(array(hugeint),array(hugeint)) -> array(hugeint)
W20230906 02:07:11.395814 537432 ExpressionFuzzer.cpp:457] Skipping unsupported signature with generic: array_has_duplicates(array(hugeint)) -> boolean
W20230906 02:07:11.397225 537432 ExpressionFuzzer.cpp:457] Skipping unsupported signature with generic: array_remove(array(hugeint),hugeint) -> array(hugeint)
W20230906 02:07:11.404963 537432 ExpressionFuzzer.cpp:457] Skipping unsupported signature with generic: trim_array(array(hugeint),bigint) -> array(hugeint)
W20230906 02:07:11.406005 537432 ExpressionFuzzer.cpp:469] Skipping non-deterministic function: rand(integer) -> integer
W20230906 02:07:11.406030 537432 ExpressionFuzzer.cpp:469] Skipping non-deterministic function: rand(smallint) -> smallint
W20230906 02:07:11.406049 537432 ExpressionFuzzer.cpp:469] Skipping non-deterministic function: rand(bigint) -> bigint
W20230906 02:07:11.406066 537432 ExpressionFuzzer.cpp:469] Skipping non-deterministic function: rand(tinyint) -> tinyint

$ tail test.log
I20230906 03:00:31.013603 537432 ExpressionFuzzer.cpp:1131] split 177 0.05% 4920
I20230906 03:00:31.013607 537432 ExpressionFuzzer.cpp:1131] combinations 176 0.05% 6199
I20230906 03:00:31.013612 537432 ExpressionFuzzer.cpp:1131] split_to_map 99 0.03% 1837
I20230906 03:00:31.013615 537432 ExpressionFuzzer.cpp:1131] array_frequency 93 0.03% 1547
I20230906 03:00:31.013619 537432 ExpressionFuzzer.cpp:1131] json_extract 31 0.01% 895
I20230906 03:00:31.013624 537432 ExpressionFuzzer.cpp:1131] json_parse 27 0.01% 1611
I20230906 03:00:31.013628 537432 ExpressionFuzzer.cpp:1131] empty_approx_set 5 0.00% 0
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

File: velox/functions/prestosql/tests/ArrayContainsTest.cpp (hunk 6 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayContainsTest.cpp]: Yes, it would generate ambiguous calls. Fixed it by using testContains and testContainsCosntantKey.

File: velox/functions/prestosql/tests/ArrayContainsTest.cpp (hunk 5 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayContainsTest.cpp]: We do not need it. Fixed it by using evaluate directly.

File: velox/functions/prestosql/tests/ArrayContainsTest.cpp (hunk 4 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayContainsTest.cpp]: Fixed

File: velox/functions/prestosql/tests/ArrayContainsTest.cpp (hunk 3 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayContainsTest.cpp]: Fixed

File: velox/functions/prestosql/ArrayContains.cpp (hunk 2 above)
Comment (duanmeng) [reply velox/functions/prestosql/ArrayContains.cpp:109]: The elements may be dictionary/constant encoded, for example, an array vector with elements from a dictionary-wrapped map vector in the test case.

```C++
// ArrayVector with DictionaryVector<Array> elements.
  auto baseVector =
      makeArrayVector<int64_t>({{1, 2, 3, 4}, {3, 4, 5}, {10, 9, 8, 7}});
  auto baseVectorSize = baseVector->size();
  const vector_size_t kTopLevelVectorSize = baseVectorSize * 2;
  BufferPtr indices =
      AlignedBuffer::allocate<vector_size_t>(kTopLevelVectorSize, pool_.get());
  auto rawIndices = indices->asMutable<vector_size_t>();
  for (size_t i = 0; i < kTopLevelVectorSize; ++i) {
    rawIndices[i] = i % baseVectorSize;
  }
  auto dictVector = BaseVector::wrapInDictionary(
      nullptr, indices, kTopLevelVectorSize, baseVector);
  auto arrayVector = makeArrayVector({0, baseVectorSize + 1}, dictVector);
```

File: velox/functions/lib/SubscriptUtil.h (hunk 1 above)
Comment (duanmeng) [reply velox/functions/lib/SubscriptUtil.h:437]: This should be the same reason as I explained in your comments in `ArrayContains.cpp`.

File: velox/functions/prestosql/ArrayContains.cpp (hunk 2 above)
Comment (laithsakka) [reply velox/functions/prestosql/ArrayContains.cpp:109]: either would have worked as explained in the comment above, however i do not have string opinion so you are free to select what you feel more suitable

File: velox/functions/prestosql/ArrayPosition.cpp [hunk 7]
```diff
@@ -137,8 +138,9 @@ void applyTypedFirstMatch(
 
     int i;
     for (i = 0; i < size; i++) {
-      if (!elementsBase->isNullAt(offset + i) &&
-          elementsBase->equalValueAt(searchBase, offset + i, searchIndex)) {
+      if (!elementsDecoded.isNullAt(offset + i) &&
+          elementsBase->equalValueAt(
```
Comment (laithsakka) [inline velox/functions/prestosql/ArrayPosition.cpp:142]: nit: it would be also nice to define equalValueAt for decoded functions

Comment (duanmeng) [reply velox/functions/prestosql/ArrayPosition.cpp:142]: OK, got it. Should we add it in a separate PR or in this one?

Comment (duanmeng) [reply velox/functions/prestosql/ArrayPosition.cpp:142]: Add a wrapper method `DecodeVector::equalValueAt`, which uses `BaseVector::equalValueAt` and `DecodeVector::indices` to compare their base vectors. Is this solution appropriate?

Comment (mbasmanova) [reply velox/functions/prestosql/ArrayPosition.cpp:142]: CC: @bikramSingh91

File: velox/functions/prestosql/tests/ArrayPositionTest.cpp [hunk 8]
```diff
@@ -834,4 +857,51 @@ TEST_F(ArrayPositionTest, invalidInstance) {
...
+  testPosition(arrayVector, {1, 2, 3, 4}, {2, 5, kTopLevelVectorSize - 7}, -1);
+}
+
+TEST_F(ArrayPositionTest, dictionaryEncodingElements) {
+  // ArrayVector with DictionaryVector<Array> elements.
+  auto baseVector = makeArrayVector<int64_t>(
+      {{1, 2, 3, 4}, {3, 4, 5}, {1, 2, 3, 4}, {10, 9, 8, 7}});
+  auto baseVectorSize = baseVector->size();
+  const vector_size_t kTopLevelVectorSize = baseVectorSize * 2;
+  BufferPtr indices =
```
Comment (laithsakka) [inline velox/functions/prestosql/tests/ArrayPositionTest.cpp:887]: nit: use BaseVector::allocateIndices()

File: velox/functions/prestosql/ArrayPosition.cpp (hunk 7 above)
Comment (duanmeng) [reply velox/functions/prestosql/ArrayPosition.cpp:142]: > @laithsakka @duanmeng Would it make sense to keep this PR simple and consider new APIs on DecodedVector separately?

@mbasmanova OK got it. I will add the new API on a separate pr.  cc @laithsakka

File: velox/functions/prestosql/tests/ArrayContainsTest.cpp [hunk 9]
```diff
@@ -108,25 +116,16 @@ TEST_F(ArrayContainsTest, varcharNoNulls) {
...
-
-    assertEqualVectors(makeNullableFlatVector<bool>(expected), result);
-  };
-
-  testContains("red", {true, false, false, true});
-  testContains("blue", {true, true, false, false});
-  testContains("yellow", {false, true, false, false});
-  testContains("green", {false, false, false, true});
-  testContains("crimson red", {false, false, false, false});
+  testContains<StringView>(arrayVector, "red", {true, false, false, true});
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayContainsTest.cpp]: Fixed

File: velox/functions/prestosql/tests/ArrayPositionTest.cpp [hunk 10]
```diff
@@ -51,6 +51,29 @@ class ArrayPositionTest : public FunctionBaseTest {
...
+      auto instanceResult = evaluate(
+          "array_position(c0, c1, c2)",
+          makeRowVector(
+              {arrayVector,
+               constSearch,
+               makeConstant(instanceOpt.value(), arrayVector->size())}));
+      assertEqualVectors(
+          makeNullableFlatVector<int64_t>(expected), instanceResult);
+    } else {
+      auto result = evaluate<SimpleVector<int64_t>>(
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayPositionTest.cpp]: Fixed

File: velox/functions/prestosql/tests/ArrayContainsTest.cpp [hunk 11]
```diff
@@ -349,4 +304,42 @@ TEST_F(ArrayContainsTest, preAllocatedNulls) {
...
+  testContainsConstantKey(arrayVector, {5, 6, 7, 8, 9}, {false, false});
+}
+
+TEST_F(ArrayContainsTest, dictionaryEncodingElements) {
+  // ArrayVector with DictionaryVector<Array> elements.
+  auto baseVector =
+      makeArrayVector<int64_t>({{1, 2, 3, 4}, {3, 4, 5}, {10, 9, 8, 7}});
+  auto baseVectorSize = baseVector->size();
+  const vector_size_t kTopLevelVectorSize = baseVectorSize * 2;
+  BufferPtr indices =
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayContainsTest.cpp]: Fixed

File: velox/functions/prestosql/tests/ArrayPositionTest.cpp (hunk 8 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayPositionTest.cpp]: Fixed

File: velox/functions/prestosql/ArrayPosition.cpp (hunk 7 above)
Comment (bikramSingh91) [reply velox/functions/prestosql/ArrayPosition.cpp:142]: @laithsakka @duanmeng Just FYI, we can also use DecodeVector::index(offset + i) instead of generating elementIndices since we will end up generating indices for identity and constant encodings which is redundant overhead. Alternatively, we can decide to handle these cases separately (have an if condition for each and only use elementIndices for dictionary vectors). Adding and using an API like DecodeVector::equalValueAt() will equate to the first option so looking forward to that follow up change. @duanmeng can you please make sure to update these callsites when you add the API?

Comment (duanmeng) [reply velox/functions/prestosql/ArrayPosition.cpp:142]: @bikramSingh91 OK got it.

**Comment by duanmeng** [comment]
@laithsakka I've rebased and pulled. All ci checks passed, could you please help to reimport?
